core/src/duplicate_shred.rs

```diff
@@ -0,0 +1,338 @@
+    let data = encoder(data).map_err(Error::EncodingError)?;
+    let chunk_size = if DUPLICATE_SHRED_HEADER_SIZE < max_size {
+        max_size - DUPLICATE_SHRED_HEADER_SIZE
+    } else {
+        return Err(Error::InvalidSizeLimit);
+    };
+    let chunks: Vec<_> = data.chunks(chunk_size).map(Vec::from).collect();
+    let num_chunks = u8::try_from(chunks.len()).map_err(|_| Error::ValueOutOfBounds)?;
+    let chunks = chunks
+        .into_iter()
+        .enumerate()
+        .map(move |(i, chunk)| DuplicateShred {
+            from: self_pubkey,
+            wallclock,
+            slot,
+            shred_index,
+            shred_type,
+            num_chunks,
+            chunk_index: i as u8,
+            chunk,
+        });
+    Ok(chunks)
+}
+
+// Returns a predicate checking if a duplicate-shred chunk matches
+// (slot, shred_index, shred_type) and has valid chunk_index.
+fn check_chunk(
+    slot: Slot,
+    shred_index: u32,
+    shred_type: ShredType,
+    num_chunks: u8,
+) -> impl Fn(&DuplicateShred) -> Result<(), Error> {
+    move |dup| {
+        if dup.slot != slot {
+            Err(Error::SlotMismatch)
+        } else if dup.shred_index != shred_index {
+            Err(Error::ShredIndexMismatch)
+        } else if dup.shred_type != shred_type {
+            Err(Error::ShredTypeMismatch)
+        } else if dup.num_chunks != num_chunks {
+            Err(Error::NumChunksMismatch)
+        } else if dup.chunk_index >= num_chunks {
+            Err(Error::InvalidChunkIndex)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+/// Reconstructs the duplicate shreds from chunks of DuplicateShred.
+pub fn into_shreds<F, I>(chunks: I, decoder: F) -> Result<(Shred, Shred), Error>
+where
+    I: IntoIterator<Item = DuplicateShred>,
+    F: FnOnce(Vec<u8>) -> Result<Vec<u8>, std::io::Error>,
+{
+    let mut chunks = chunks.into_iter();
+    let DuplicateShred {
+        slot,
+        shred_index,
+        shred_type,
+        num_chunks,
+        chunk_index,
+        chunk,
+        ..
+    } = match chunks.next() {
+        None => return Err(Error::InvalidDuplicateShreds),
+        Some(chunk) => chunk,
+    };
+    let check_chunk = check_chunk(slot, shred_index, shred_type, num_chunks);
+    let mut data = HashMap::new();
+    data.insert(chunk_index, chunk);
+    for chunk in chunks {
+        check_chunk(&chunk)?;
+        match data.entry(chunk.chunk_index) {
+            Entry::Vacant(entry) => {
+                entry.insert(chunk.chunk);
+            }
+            Entry::Occupied(entry) => {
+                if *entry.get() != chunk.chunk {
+                    return Err(Error::DataChunkMismatch);
+                }
+            }
+        }
+    }
+    if data.len() != num_chunks as usize {
+        return Err(Error::MissingDataChunk);
+    }
+    let data = (0..num_chunks).map(|k| data.remove(&k).unwrap());
+    let data = decoder(data.concat()).map_err(Error::DecodingError)?;
```

Comment:
Would like to avoid `map_err` if possible, see: https://github.com/solana-labs/solana/pull/14133/files#r544905497

Reply:
ah I missed that it was already wrapped, that works!